orc/src/main/java/org/apache/iceberg/orc/ORC.java

```diff
@@ -223,6 +232,161 @@ public DataWriteBuilder withSortOrder(SortOrder newSortOrder) {
+      this.partition = key;
+      return this;
+    }
+
+    public DeleteWriteBuilder withKeyMetadata(EncryptionKeyMetadata metadata) {
+      this.keyMetadata = metadata;
+      return this;
+    }
+
+    public DeleteWriteBuilder equalityFieldIds(List<Integer> fieldIds) {
+      this.equalityFieldIds = ArrayUtil.toIntArray(fieldIds);
+      return this;
+    }
+
+    public DeleteWriteBuilder equalityFieldIds(int... fieldIds) {
+      this.equalityFieldIds = fieldIds;
+      return this;
+    }
+
+    public DeleteWriteBuilder transformPaths(Function<CharSequence, ?> newPathTransformFunc) {
+      this.pathTransformFunc = newPathTransformFunc;
+      return this;
+    }
+
+    public DeleteWriteBuilder withSortOrder(SortOrder newSortOrder) {
+      this.sortOrder = newSortOrder;
+      return this;
+    }
+
+    public <T> EqualityDeleteWriter<T> buildEqualityWriter() throws IOException {
+      Preconditions.checkState(rowSchema != null, "Cannot create equality delete file without a schema`");
+      Preconditions.checkState(equalityFieldIds != null, "Cannot create equality delete file without delete field ids");
+      Preconditions.checkState(createWriterFunc != null,
+          "Cannot create equality delete file unless createWriterFunc is set");
+      Preconditions.checkArgument(spec != null,
+          "Spec must not be null when creating equality delete writer");
+      Preconditions.checkArgument(spec.isUnpartitioned() || partition != null,
+          "Partition must not be null for partitioned writes");
+
+      meta("delete-type", "equality");
+      meta("delete-field-ids", IntStream.of(equalityFieldIds)
+          .mapToObj(Objects::toString)
+          .collect(Collectors.joining(", ")));
+
+      // the appender uses the row schema without extra columns
+      appenderBuilder.schema(rowSchema);
+      appenderBuilder.createWriterFunc(createWriterFunc);
+
+      return new EqualityDeleteWriter<>(
+          appenderBuilder.build(), FileFormat.ORC, location, spec, partition, keyMetadata,
+          sortOrder, equalityFieldIds);
+    }
+
+    public <T> PositionDeleteWriter<T> buildPositionWriter() throws IOException {
+      Preconditions.checkState(equalityFieldIds == null, "Cannot create position delete file using delete field ids");
+      Preconditions.checkArgument(spec != null,
+          "Spec must not be null when creating position delete writer");
+      Preconditions.checkArgument(spec.isUnpartitioned() || partition != null,
+          "Partition must not be null for partitioned writes");
+
+      meta("delete-type", "position");
+
+      Schema deleteSchema =
+          rowSchema == null ? DeleteSchemaUtil.pathPosSchema() : DeleteSchemaUtil.posDeleteSchema(rowSchema);
+      appenderBuilder.schema(deleteSchema);
+
+      if (createWriterFunc != null) {
+        appenderBuilder.createWriterFunc((schema, typeDescription) -> {
+          OrcRowWriter<?> writer = createWriterFunc.apply(deleteSchema, typeDescription);
```

Comment:
I am a bit confused here. I'd probably align the design and logic with `PositionDeleteStructWriter` in `Parquet`.

Reply:
I was a bit confused about the `Parquet` implementation, and I might not understand it fully yet.
In `Parquet` we have this code:
```
        appenderBuilder.createWriterFunc(parquetSchema -> {
          ParquetValueWriter<?> writer = createWriterFunc.apply(parquetSchema);
          if (writer instanceof StructWriter) {
            return new PositionDeleteStructWriter<T>((StructWriter<?>) writer, pathTransformFunc);
          } else {
            throw new UnsupportedOperationException("Cannot wrap writer for position deletes: " + writer.getClass());
          }
        });
```

My understanding is that we create the supplied writer here:
```
          ParquetValueWriter<?> writer = createWriterFunc.apply(parquetSchema);
```

Then we check if it is the correct type, and if so we create the `PositionDeleteStructWriter` where the internal writer is replaced by the one generated above:
```
            return new PositionDeleteStructWriter<T>((StructWriter<?>) writer, pathTransformFunc);
```

I might miss something, but this is what I have tried to replicate with `ORC` as well.
Created the internal writer:
```
          OrcRowWriter<?> writer = createWriterFunc.apply(deleteSchema, typeDescription);
```
No need to check the type here, as the `createWriterFunc` ensures it. So we can go ahead to create a wrapper writer:
```
          return GenericOrcWriters.positionDelete(deleteSchema, writer);
```

The only difference I see is that we do not handle the `pathTransformFunc`. Do I miss something?

Thanks,
Peter
```

Reply:
@pvary @aokolnychyi ,   The way why we need to introduce a `pathTransformFunc` is because the parquet are exposing API to load columns by column description rather than column id,  I think this [comment](https://github.com/apache/iceberg/pull/1836/files#r531474478) will provide the fully background.

Reply:
I think ORC can use the same approach as AVRO to generate the PositionAndRowDatumWriter because we ORC don't need the column description to locate the specific column.